### ISSUE TITLE
[question] How to get environment variables from profile.

### ISSUE THREAD

**Issue by skycaptain** [conan-io/conan#13545]
### What is your question?

We use the Yocto SDK to cross-compile for embedded devices. The SDK ships Qemu, which allows us to run unittests and `test_package` executables on the build host. We set an environment variable `QEMU` in the `[runenv]` section of our profile, which contains the path to the qemu binary. In our Conan recipe we use the following test function:

```
def test(self):
        cmd = os.path.join(self.cpp.build.bindirs[0], "my-lib-a-test-package")

        # FIXME: Is there an easier way to get runenv variables?
        runenvs = dict(self.runenv.vars(self, scope="run").items())

        if not can_run(self) and "QEMU" in runenvs:
            cmd = "{QEMU} {cmd}".format(**runenvs, cmd=cmd)

        self.run(cmd, env="conanrun")
```

My question is how we can get access to runenv variables. It seems like `Environment` stores variables as an `OrderedDict` internally, the same is true for `EnvVars`. However, both classes to not implement `collections.Container`, i.e. do not implement `__contains__`. Is this intentional? Would you accept PRs to add this functionality?

(We would also just rely on the shell and use `self.run(f"${{QEMU}} {cmd}", env="conanrun")` to interpolate the variables in shel; however we prefer to have the full path also in the logs generated by Conan, hence interpolation in the Conan recipe is preferred)

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @skycaptain 

Thanks for your question, it is an interesting issue.
It might be possible to implement ``__contains__`` in ``Environment`` or in ``EnvVars``, seems easy, so sure, feel free to submit a PR (to ``release/2.0`` branch)

In any case, in the meantime maybe something like this is easier?:
```python
def test(self):
        cmd = os.path.join(self.cpp.build.bindirs[0], "my-lib-a-test-package")
        qemu = self.runenv.vars(self, scope="run").get("QEMU")

        if not can_run(self) and qemu:
            cmd = "{qemu} {cmd}".format(qemu=qemu, cmd=cmd)

        self.run(cmd, env="conanrun")
```
I mean, this is kind of more pythonic approach, if you expect to use the value, do not check if ``thing in container``, but directly ask for it, and use it if not None.

**Comment by skycaptain** [comment]
Thank you that's way better. I will submit a PR; `Environment` and `EnvVars` behaving like a dict (they already have `__getitem__`, `get` and `items`) would even be more pythonic.

**Comment by skycaptain** [comment]
Upon closer inspection, it appears that the above code fragment is not as straightforward as it may seem. While it works if the variables are defined in the `[runenv]` section of the profile, it does not work when `QEMU` is defined with `runenv_info` in a `tool_requires` package. In this case, `self.runenv.vars` is empty when `test()` of a `test_package` is executed. It seems that this issue has already been reported here: https://github.com/conan-io/conan/issues/13536. This seems inconsistent to me.

**Comment by memsharded** [comment]
Yes, that is true, the above was suggested as a solution that works when defined in the profile. But if you move it to a package it will not work.

If the env-var can come from different locations, then indeed the ``VirtualRunEnv`` would probably be the best to get it, so something like this:

```python
def generate(self):
     env = VirtualRunEnv(self)
     envvars= env.vars(scope="run")
     qemu = envvars.get("QEMU")
     save(self, "myqemufile", qemu or "")
     env.generate()

def test(self):
      qemu = load(self, "myqemufile")
      # use qemu
```

This is the general approach: ``generate()`` collects information from dependencies, profiles, etc, and prepares the build, creating the necessary files. Then the ``build()`` or ``test()`` will use the processed information accordingly.

This is the way that a user can work and reproduce locally, ``conan install`` and it will get the environment files and other stuff that they need to do the build.

**Comment by skycaptain** [comment]
Yes, we just had the same thought. However, it seems that VirtualRunEnv only contains `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH`:

```
my-lib-a/0.1.0 (test package): {'LD_LIBRARY_PATH': '/home/vscode/.conan2/p/my-dca7fcddd660a3/p/lib', 'DYLD_LIBRARY_PATH': '/home/vscode/.conan2/p/my-dca7fcddd660a3/p/lib'}
```

`QEMU` is only in `VirtualBuildEnv`. I've double-checked, to be sure we add it to `self.runenv_info` in our test_requires package:

```
my-sdk/0.1.0: Defining run env: QEMU=qemu-aarch64 -L /opt/my-sdk/sysroots/cortexa57-dpg-linux -E LD_LIBRARY_PATH=/opt/my-sdk/sysroots/cortexa57-dpg-linux/usr/lib:/opt/my-sdk/sysroots/cortexa57-dpg-linux/lib
```

`QEMU` is added to `VirtualRunEnv` if `my-sdk` is also added as `test_requires`.

**Comment by memsharded** [comment]
I think that is expected, depending how you consider things:
- VirtualRunEnv is expected to be the "host" running environment, and contain only "host" dependencies. If you are cross building from Windows to Linux, it should only contain things that can run in Linux
- VirtualBuildEnv is expected to be the "build" running environment and contain information for build time. For executables that need to run, they should definitely be "build" dependencies, though it is true that also some "host" regular ``requires`` can also provide "build-env" information for build time too.

Maybe the thing is that ``qemu`` kind of lies in between? I'd need to understand a bit better about the setup of the ``my-sdk`` and how it is expected to be used, who is doing the ``test_requires`` to the ``my-sdk`` to depend on ``qemu``, the ``test_package/conanfile.py``?

**Comment by memsharded** [comment]
Hi @skycaptain 

It seems this ticket went staled and not followed up after my last comment.
Marking it as staled to be closed soon if no further activity.

If you'd like to keep the conversation going, just let us know. Many thanks!

**Comment by skycaptain** [comment]
Hey @memsharded. Many thanks for the reminder. I switched companies and positions and meanwhile I'm not working with Conan anymore. Hence closing.

**Comment by memsharded** [comment]
Thanks for the feedback!
